velox/functions/sparksql/ToPrettyString.h

```diff
@@ -0,0 +1,206 @@
+          } catch (const std::exception& e) {
+            return Status::Invalid(e.what());
+          }
+          return Status::OK();
+        }
+      }
+      const auto castResult =
+          util::Converter<TypeKind::VARCHAR, void, util::SparkCastPolicy>::
+              tryCast(*input);
+      VELOX_DCHECK(!castResult.hasError());
+      result.copy_from(castResult.value());
+    } else {
+      result.setNoCopy(detail::kNull);
+    }
+    return Status::OK();
+  }
+
+ private:
+  TypePtr inputType_;
+};
+
+/// Returns pretty string for varbinary. It has several differences with
+/// cast(varbinary as string):
+/// 1) It prints null input as "NULL" rather than producing null output.
+/// 2) It prints binary value as hex string representation rather than UTF-8.
+/// The pretty string is composed of the hex digits of bytes and spaces between
+/// them. E.g., the result of to_pretty_string("abc") is "[31 32 33]".
+template <typename TExec>
+struct ToPrettyStringVarbinaryFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  template <typename TInput>
+  void callNullable(out_type<Varchar>& result, const TInput* input) {
+    if (input) {
+      // One byte spares 2 char, and with the spaces and the boxes.
+      // Byte size: 2 * input->size(), spaces size: input->size() - 1, boxes
+      // size: 2, its sum is 1 + 3 * input->size().
+      result.resize(1 + 3 * input->size());
+      char* const startPosition = result.data();
+      char* pos = startPosition;
+      *pos++ = '[';
+      for (auto i = 0; i < input->size(); i++) {
+        int count = std::sprintf(pos, "%02X", input->data()[i]);
+        VELOX_DCHECK_EQ(count, 2);
+        pos += 2;
+        *pos++ = ' ';
+      }
+      *--pos = ']';
+    } else {
+      result.setNoCopy(detail::kNull);
+    }
+  }
+};
+
+/// Returns pretty string for Timestamp. It has one difference with
+/// cast(timestamp as string):
+/// 1) It prints null input as "NULL" rather than producing null output.
+template <typename TExec>
+struct ToPrettyStringTimestampFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  void initialize(
+      const std::vector<TypePtr>& /*inputTypes*/,
+      const core::QueryConfig& config,
+      const arg_type<Timestamp>* /*timestamp*/) {
+    auto timezone = config.sessionTimezone();
+    if (!timezone.empty()) {
+      options_.timeZone = tz::locateZone(timezone);
+    }
+    timestampRowSize_ = getMaxStringLength(options_);
+  }
+
+  Status callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Timestamp>* timestamp) {
+    if (timestamp) {
+      Timestamp inputValue(*timestamp);
+      try {
+        if (options_.timeZone) {
+          inputValue.toTimezone(*(options_.timeZone));
+        }
+        result.reserve(timestampRowSize_);
+        const auto stringView =
+            Timestamp::tsToStringView(inputValue, options_, result.data());
+        result.resize(stringView.size());
+      } catch (const std::exception& e) {
```

Comment:
`CastExpr::applyToSelectedNoThrowLocal` adopts the below logic for NULL.

https://github.com/facebookincubator/velox/blob/ccd765d3d9e70ddc4cb222b00433b0922915b6cc/velox/expression/CastExpr-inl.h#L206-L214

Reply:
Do you mean match the VeloxException, then catch std::exception? @rui-mo This is not addressed. @ArnavBalyan

Reply:
Current implementation catches all the std::exception and converts them as user error, while in the cast, exception is thrown if not a user error, and set null result if it is. If the only differences with cast are the listed ones, we might align with cast here.